spec/index.bs

```diff
@@ -873,6 +927,103 @@ interface GPUBufferUsage {
 
 ## Buffer Mapping ## {#buffer-mapping}
 
+An application can request to map a {{GPUBuffer}} to get its mapping which is
+an {{ArrayBuffer}} representing the {{GPUBuffer}}'s allocation. Mappings are
+requested asynchronously so that the user agent can ensure the GPU is finished
+using the {{GPUBuffer}} before the application gets its mapping. Mappings can
+be requested for reading with {{GPUBuffer/mapReadAsync}} or writing with
+{{GPUBuffer/mapWriteAsync}}. A mapped {{GPUBuffer}} cannot be used by the GPU
+and must be unmapped using {{GPUBuffer/unmap}} before it can be used by the GPU
+again.
+
+<!-- TODO(kangz) add client-side validation that a mapped buffer can only be destroyed on the worker on which it was mapped -->
+
+### {{GPUBuffer/mapReadAsync|GPUDevice.mapReadAsync}} ### {#GPUBuffer-mapReadAsync}
+
+<dl dfn-type="method" dfn-for="GPUBuffer">
+    : <dfn>mapReadAsync()</dfn>
+    ::
+        <div algorithm="GPUBuffer.mapReadAsync()">
+            <!-- TODO(kangz) handle error buffers once we have a description of the error monad -->
+            1. If the {{[[usage]]}} slot of |this| doesn't contain the {{GPUBufferUsage/MAP_READ}} bit:
+
+                1. Record a validation error on the current scope.
+                1. Return a new promise that rejects. <!-- TODO(kangz) specify on that it happens on the device timeline-->
+
+            1. Let |p| be a new {{Promise}}.
+            1. Set the {{[[mapping_promise]]}} slot of |this| to |p|.
+            1. Set the {{[[state]]}} slot of |this| to [=buffer state/mapping pending for reading=].
+            1. Enqueue an operation on the [=Queue timeline=] that will execute the following:
+
+                1. Let |m| be a new {{ArrayBuffer}} of size the {{[[size]]}} of |this|.
+                1. Set the content of |m| to the content of |this|'s allocation.
+                1. Set the {{[[state]]}} slot of |this| to [=buffer state/mapped for reading=].
+                1. If |p| is pending:
```

Comment:
in what case would `p` not be pending?

Reply:
huh, that opens a can of works for me. So the algorithms we describe here, they appear to be conceptual, and they are technically spread over different timelines. At the spec level, this is a situation where one algorithm (for `unmap()`) is executed in the middle of another algorithm (for `mapReadAsync()`)... yikes!

Is the reader supposed to assume complete lack of synchronization between any algorithms defined here in the spec? It seems to me that we'll easily find more cases where running one algorithm in the middle of another would lead to inconsistent state. This isn't an *actual* problem to the implementations, because they have internal synchronization at various pipeline stages.

Seems non-trivial to address that in the spec though. Only thing comes to mind is splitting the algorithm description on the timelines we have defined and clearly stating somewhere that *between* the timelines (i.e. after the last command on timeline X and before the first command on timeline Y) some time may pass, and other algorithms may be executing.

Anyhow, sounds like a big topic to me, would like to raise the issue on the next call.

Reply:
Notice that all of this section of the algorithm is nested under a "1. Enqueue an operation on the [=Queue timeline=] that will execute the following:". So this part of the algorithm acts like a closure that's executed at some point on the Queue timeline. We could have a more explicit callout of this concept in the timeline section that this algorithm could refer to.

Happy to talk about it on the next call.